configparser breaks if configuration is in path containing %

* papis version (`$ papis --version` or commit number): HEAD (b6fe8962)

# how to reproduce the issue
Clone `papis` project under a path containing a `%` character.
Try running `pytest`, result:

```
================================================= ERRORS =================================================
____________________________________ ERROR at setup of test_edit_run _____________________________________

request = <SubRequest 'tmp_library' for <Function test_edit_run>>

    @pytest.fixture(scope="function")
    def tmp_library(request: SubRequest) -> Iterator[TemporaryLibrary]:
        """A fixture that creates a :class:`TemporaryLibrary`.
    
        Additional keyword arguments can be passed using the ``library_setup`` marker
    
        .. code:: python
    
            @pytest.mark.library_setup(use_git=False)
            def test_me(tmp_library: TemporaryLibrary) -> None:
                ...
        """
        marker = request.node.get_closest_marker("library_setup")
        kwargs = marker.kwargs if marker else {}
    
        # NOTE: support indirect fixture parametrizations that overwrite markers
        kwargs.update(getattr(request, "param", {}))
    
>       with TemporaryLibrary(**kwargs) as lib:

papis/testing.py:588: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
papis/testing.py:344: in __enter__
    super().__enter__()
papis/testing.py:279: in __enter__
    config.read_dict(settings)
/usr/lib/python3.12/configparser.py:741: in read_dict
    self.set(section, key, value)
/usr/lib/python3.12/configparser.py:1184: in set
    super().set(section, option, value)
/usr/lib/python3.12/configparser.py:872: in set
    value = self._interpolation.before_set(self, section, option,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <configparser.BasicInterpolation object at 0x71607afb0f80>
parser = <configparser.ConfigParser object at 0x716075c35fd0>, section = 'settings', option = 'editor'
value = '/usr/bin/python /home/gesh/aur/1%mine/python/papis-git/src/papis/tests/commands/scripts.py sed'

    def before_set(self, parser, section, option, value):
        tmp_value = value.replace('%%', '') # escaped percent signs
        tmp_value = self._KEYCRE.sub('', tmp_value) # valid syntax
        if '%' in tmp_value:
>           raise ValueError("invalid interpolation syntax in %r at "
                             "position %d" % (value, tmp_value.find('%')))
E           ValueError: invalid interpolation syntax in '/usr/bin/python /home/gesh/aur/1%mine/python/papis-git/src/papis/tests/commands/scripts.py sed' at position 32

/usr/lib/python3.12/configparser.py:374: ValueError
____________________________________ ERROR at setup of test_edit_cli _____________________________________

request = <SubRequest 'tmp_library' for <Function test_edit_cli>>

    @pytest.fixture(scope="function")
    def tmp_library(request: SubRequest) -> Iterator[TemporaryLibrary]:
        """A fixture that creates a :class:`TemporaryLibrary`.
    
        Additional keyword arguments can be passed using the ``library_setup`` marker
    
        .. code:: python
    
            @pytest.mark.library_setup(use_git=False)
            def test_me(tmp_library: TemporaryLibrary) -> None:
                ...
        """
        marker = request.node.get_closest_marker("library_setup")
        kwargs = marker.kwargs if marker else {}
    
        # NOTE: support indirect fixture parametrizations that overwrite markers
        kwargs.update(getattr(request, "param", {}))
    
>       with TemporaryLibrary(**kwargs) as lib:

papis/testing.py:588: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
papis/testing.py:344: in __enter__
    super().__enter__()
papis/testing.py:279: in __enter__
    config.read_dict(settings)
/usr/lib/python3.12/configparser.py:741: in read_dict
    self.set(section, key, value)
/usr/lib/python3.12/configparser.py:1184: in set
    super().set(section, option, value)
/usr/lib/python3.12/configparser.py:872: in set
    value = self._interpolation.before_set(self, section, option,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <configparser.BasicInterpolation object at 0x71607afb0f80>
parser = <configparser.ConfigParser object at 0x716075c627b0>, section = 'settings', option = 'editor'
value = '/usr/bin/python /home/gesh/aur/1%mine/python/papis-git/src/papis/tests/commands/scripts.py echo'

    def before_set(self, parser, section, option, value):
        tmp_value = value.replace('%%', '') # escaped percent signs
        tmp_value = self._KEYCRE.sub('', tmp_value) # valid syntax
        if '%' in tmp_value:
>           raise ValueError("invalid interpolation syntax in %r at "
                             "position %d" % (value, tmp_value.find('%')))
E           ValueError: invalid interpolation syntax in '/usr/bin/python /home/gesh/aur/1%mine/python/papis-git/src/papis/tests/commands/scripts.py echo' at position 32

/usr/lib/python3.12/configparser.py:374: ValueError
____________________________________ ERROR at setup of test_open_cli _____________________________________

request = <SubRequest 'tmp_library' for <Function test_open_cli>>

    @pytest.fixture(scope="function")
    def tmp_library(request: SubRequest) -> Iterator[TemporaryLibrary]:
        """A fixture that creates a :class:`TemporaryLibrary`.
    
        Additional keyword arguments can be passed using the ``library_setup`` marker
    
        .. code:: python
    
            @pytest.mark.library_setup(use_git=False)
            def test_me(tmp_library: TemporaryLibrary) -> None:
                ...
        """
        marker = request.node.get_closest_marker("library_setup")
        kwargs = marker.kwargs if marker else {}
    
        # NOTE: support indirect fixture parametrizations that overwrite markers
        kwargs.update(getattr(request, "param", {}))
    
>       with TemporaryLibrary(**kwargs) as lib:

papis/testing.py:588: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
papis/testing.py:344: in __enter__
    super().__enter__()
papis/testing.py:279: in __enter__
    config.read_dict(settings)
/usr/lib/python3.12/configparser.py:741: in read_dict
    self.set(section, key, value)
/usr/lib/python3.12/configparser.py:1184: in set
    super().set(section, option, value)
/usr/lib/python3.12/configparser.py:872: in set
    value = self._interpolation.before_set(self, section, option,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <configparser.BasicInterpolation object at 0x71607afb0f80>
parser = <configparser.ConfigParser object at 0x7160758d86b0>, section = 'settings'
option = 'file-browser'
value = '/usr/bin/python /home/gesh/aur/1%mine/python/papis-git/src/papis/tests/commands/scripts.py echo'

    def before_set(self, parser, section, option, value):
        tmp_value = value.replace('%%', '') # escaped percent signs
        tmp_value = self._KEYCRE.sub('', tmp_value) # valid syntax
        if '%' in tmp_value:
>           raise ValueError("invalid interpolation syntax in %r at "
                             "position %d" % (value, tmp_value.find('%')))
E           ValueError: invalid interpolation syntax in '/usr/bin/python /home/gesh/aur/1%mine/python/papis-git/src/papis/tests/commands/scripts.py echo' at position 32

/usr/lib/python3.12/configparser.py:374: ValueError
```

Not necessarily something that needs fixing, but at least could be worth a
warning in `HACKING.md` or something. (and means I need to find some _other_
intra-filename separator, after `#` and `:` have both caused trouble, cf eg
https://github.com/CycloneDX/cyclonedx-python-lib/issues/551)

--- a/doc/source/configuration.rst
+++ b/doc/source/configuration.rst
@@ -155,6 +155,26 @@ all the options and more extensive descriptions)
     browse-query-format = {doc[first_name]} {doc[last_name]}
     add-open = False
 
+.. note::
+
+   Papis uses the standard :class:`~configparser.ConfigParser` to read and write
+   configuration options. It also allows interpolation of values
+   (using :class:`~configparser.BasicInterpolation`) so that previous values
+   can be referred back to. For example, one can do
+
+   .. code:: ini
+
+        [DEFAULT]
+        basedir = /path/to/libraries
+
+        [papers]
+        dir = %(basedir)s/papers
+
+        [books]
+        dir = %(basedir)s/books
+
+   Interpolation variables are looked for in the current section and in the
+   standard ``DEFAULT`` section.
 
 Local configuration files
 -------------------------

--- a/papis/commands/addto.py
+++ b/papis/commands/addto.py
@@ -162,7 +162,7 @@ def cli(query: str,
     document = docs[0]
 
     if file_name is not None:  # Use args if set
-        papis.config.set("add-file-name", file_name)
+        papis.config.set("add-file-name", papis.config.escape_interp(file_name))
 
     try:
         run(document, files + urls, git=git, link=link)

--- a/papis/commands/edit.py
+++ b/papis/commands/edit.py
@@ -108,7 +108,7 @@ def cli(query: str,
         return
 
     if editor is not None:
-        papis.config.set("editor", editor)
+        papis.config.set("editor", papis.config.escape_interp(editor))
 
     for document in documents:
         if notes:

--- a/papis/commands/open.py
+++ b/papis/commands/open.py
@@ -95,7 +95,7 @@ def run(document: papis.document.Document,
         folder: bool = False,
         mark: bool = False) -> None:
     if opener is not None:
-        papis.config.set("opentool", opener)
+        papis.config.set("opentool", papis.config.escape_interp(opener))
 
     _doc_folder = document.get_main_folder()
     if _doc_folder is None:
@@ -136,7 +136,7 @@ def run(document: papis.document.Document,
                         papis.document.from_data(mark_dict[0]),
                         doc_key=_mark_name)
                     logger.info("Setting opener to '%s'.", opener)
-                    papis.config.set("opentool", opener)
+                    papis.config.set("opentool", papis.config.escape_interp(opener))
         files = document.get_files()
         if not files:
             logger.error("The chosen document has no files attached: '%s'.",
@@ -169,7 +169,7 @@ def cli(query: str, doc_folder: Tuple[str, ...], tool: str, folder: bool,
         mark: bool) -> None:
     """Open document from a given library"""
     if tool:
-        papis.config.set("opentool", tool)
+        papis.config.set("opentool", papis.config.escape_interp(tool))
 
     documents = papis.cli.handle_doc_folder_query_all_sort(query,
                                                            doc_folder,

--- a/papis/config.py
+++ b/papis/config.py
@@ -549,7 +549,8 @@ def set_lib(library: papis.library.Library) -> None:
 
     if library.name not in config:
         # NOTE: can't use set(...) here due to cyclic dependencies
-        config[library.name] = {"dirs": str(library.paths)}
+        paths = [escape_interp(path) for path in library.paths]
+        config[library.name] = {"dirs": str(paths)}
 
     CURRENT_LIBRARY = library
 
@@ -577,7 +578,7 @@ def get_lib_from_name(libname: str) -> papis.library.Library:
 
             lib = papis.library.from_paths([libname])
             # NOTE: can't use set(...) here due to cyclic dependencies
-            config[lib.name] = {"dirs": str(lib.paths)}
+            config[lib.name] = {"dirs": str([escape_interp(libname)])}
         else:
             raise RuntimeError(
                 f"Library '{libname}' does not seem to exist. "
@@ -679,3 +680,18 @@ def reset_configuration() -> Configuration:
 
     logger.debug("Resetting configuration.")
     return get_configuration()
+
+
+def escape_interp(path: str) -> str:
+    """Escape paths added to the configuration file.
+
+    By default, the :class:`papis.config.Configuration` enables string interpolation
+    in the key values (e.g. using ``key = %(other_key)s-suffix)``). Any paths
+    added to the configuration should then be escaped so that they do not
+    interfere with the interpolation.
+    """
+    import re
+
+    # FIXME: this should be smart enough to not double quote valid interpolation
+    # paths? Would need a regex to skip things like `%\(\w+\)`?
+    return re.sub(r"([^%])%([^%(])", r"\1%%\2", path)

--- a/papis/config.py
+++ b/papis/config.py
@@ -312,7 +312,15 @@ def set(key: str, value: Any, section: Optional[str] = None) -> None:
     if not config.has_section(section):
         config.add_section(section)
 
-    config[section][key] = str(value)
+    try:
+        config[section][key] = str(value)
+    except ValueError as exc:
+        logger.error("Failed to set the key '%s' in section '%s' with value '%s'.",
+                     key, section, value)
+        logger.error("If the value contains a '%', this should be properly "
+                     "escaped (using a double percent '%%') or the proper "
+                     "interpolation syntax should be used (i.e. '%(other_key)').",
+                     exc_info=exc)
 
 
 def general_get(key: str,

--- a/papis/testing.py
+++ b/papis/testing.py
@@ -614,6 +614,8 @@ def resource_cache(request: SubRequest) -> ResourceCache:
 
 
 def pytest_addoption(parser: Parser) -> None:
+    parser.addoption("--papis-enable-logging", action="store_true",
+                     help="Enable logging while running tests")
     parser.addoption("--papis-tmp-doctests", action="store_true",
                      help="Use a temporary configuration file for doctests")
     parser.addoption("--papis-tmp-xdg-home", action="store_true",
@@ -626,6 +628,10 @@ def pytest_configure(config: Config) -> None:
         # a chance to construct a TemporaryConfiguration. We set XDG_CONFIG_HOME
         # here first to avoid them picking up any sort of user configuration.
         os.environ["XDG_CONFIG_HOME"] = tempfile.gettempdir()
+
+    if config.getoption("--papis-enable-logging"):
+        import papis.logging
+        papis.logging.setup()
 
     config.addinivalue_line(
         "markers",

--- a/papis/testing.py
+++ b/papis/testing.py
@@ -200,6 +200,7 @@ class TemporaryConfiguration:
     libname: ClassVar[str] = "test"
 
     def __init__(self,
+                 prefix: str = "papis-test-",
                  settings: Optional[Dict[str, Any]] = None,
                  overwrite: bool = False) -> None:
         #: A set of settings to be added to the configuration on creation
@@ -216,6 +217,9 @@ class TemporaryConfiguration:
         #: When entering the context manager, this will contain the config
         #: file used by papis.
         self.configfile: str = ""
+
+        #: Prefix for the temporary directory created for the test.
+        self.prefix = prefix
 
         self._tmpdir: Optional[tempfile.TemporaryDirectory[str]] = None
         self._monkeypatch: Optional[pytest.MonkeyPatch] = None
@@ -241,7 +245,7 @@ class TemporaryConfiguration:
 
         # create directories and files
         self._monkeypatch = pytest.MonkeyPatch()
-        self._tmpdir = tempfile.TemporaryDirectory(prefix="papis-test-")
+        self._tmpdir = tempfile.TemporaryDirectory(prefix=self.prefix)
 
         self.libdir = os.path.join(self.tmpdir, "lib")
         self.configdir = os.path.join(self.tmpdir, "papis")
@@ -253,8 +257,10 @@ class TemporaryConfiguration:
         os.makedirs(self.configscripts)
 
         # load settings
+        import papis.config
+
         settings = {
-            self.libname: {"dir": self.libdir},
+            self.libname: {"dir": papis.config.escape_interp(self.libdir)},
             "settings": {"default-library": self.libname}
         }
 
@@ -289,7 +295,6 @@ class TemporaryConfiguration:
         #   * reset papis.plugin managers
 
         # reload configuration
-        import papis.config
         papis.config.set_config_file(self.configfile)
         papis.config.reset_configuration()
 

--- a/tests/commands/test_edit.py
+++ b/tests/commands/test_edit.py
@@ -17,7 +17,9 @@ def get_mock_script(name: str) -> str:
 
     script = os.path.join(os.path.dirname(__file__), "scripts.py")
 
-    return "{} {} {}".format(sys.executable, script, name)
+    from papis.config import escape_interp
+
+    return escape_interp("{} {} {}".format(sys.executable, script, name))
 
 
 @pytest.mark.library_setup(settings={
@@ -66,7 +68,8 @@ def test_edit_cli(tmp_library: TemporaryLibrary) -> None:
         cli,
         ["--all", "--editor", ls, "krishnamurti"])
     assert result.exit_code == 0
-    assert papis.config.get("editor") == get_mock_script("ls")
+    assert (papis.config.escape_interp(papis.config.get("editor"))
+            == get_mock_script("ls"))
 
     # check --notes
     notes_name = papis.config.get("notes-name")
@@ -76,7 +79,8 @@ def test_edit_cli(tmp_library: TemporaryLibrary) -> None:
         cli,
         ["--all", "--editor", get_mock_script("echo"), "--notes", "krishnamurti"])
     assert result.exit_code == 0
-    assert papis.config.get("editor") == get_mock_script("echo")
+    assert (papis.config.escape_interp(papis.config.get("editor"))
+            == get_mock_script("echo"))
 
     db = papis.database.get()
     doc, = db.query_dict({"author": "Krishnamurti"})

--- a/tests/commands/test_open.py
+++ b/tests/commands/test_open.py
@@ -4,11 +4,24 @@ import pytest
 
 from papis.testing import TemporaryLibrary, PapisRunner
 
-script = os.path.join(os.path.dirname(__file__), "scripts.py")
+
+def get_mock_script(name: str) -> str:
+    """Constructs a command call for a command mocked by ``scripts.py``.
+
+    :param name: the name of the command, e.g. ``ls`` or ``echo``.
+    :returns: a string of the command
+    """
+    import sys
+
+    script = os.path.join(os.path.dirname(__file__), "scripts.py")
+
+    from papis.config import escape_interp
+
+    return escape_interp("{} {} {}".format(sys.executable, script, name))
 
 
 @pytest.mark.library_setup(settings={
-    "file-browser": "{} {} echo".format(sys.executable, script)
+    "file-browser": get_mock_script("echo")
     })
 def test_open_cli(tmp_library: TemporaryLibrary) -> None:
     from papis.commands.open import cli
@@ -35,7 +48,7 @@ def test_open_cli(tmp_library: TemporaryLibrary) -> None:
     # Use a mock scriptlet
     result = cli_runner.invoke(
         cli,
-        ["--tool", "{} {} echo".format(sys.executable, script),
+        ["--tool", get_mock_script("echo"),
          "--mark", "--all", "Krishnamurti"])
     assert result.exit_code == 0
 

--- a/tests/test_config.py
+++ b/tests/test_config.py
@@ -31,9 +31,21 @@ def test_get_config_paths(tmp_config: TemporaryConfiguration) -> None:
     assert papis.config.get_scripts_folder() == scriptsdir
 
 
-def test_get_config_home(tmp_config: TemporaryConfiguration, monkeypatch) -> None:
+def test_get_config_home(tmp_config: TemporaryConfiguration,
+                         monkeypatch: pytest.MonkeyPatch) -> None:
     import papis.config
     assert re.match(r".+papis", papis.config.get_config_home()) is not None
+
+
+def test_config_interpolation() -> None:
+    with TemporaryConfiguration(prefix="papis%test%") as tmp_config:
+        import papis.config
+
+        assert papis.config.get("dir", section=tmp_config.libname) == tmp_config.libdir
+
+        papis.config.set("some_value", "value1")
+        papis.config.set("more_value", "more_%(some_value)s")
+        assert papis.config.get("more_value") == "more_value1"
 
 
 def test_set(tmp_config: TemporaryConfiguration) -> None:
@@ -119,7 +131,8 @@ def test_get_types(tmp_config: TemporaryConfiguration) -> None:
         value = papis.config.getboolean("boolean_config")
 
 
-def test_get_configuration(tmp_config: TemporaryConfiguration, monkeypatch) -> None:
+def test_get_configuration(tmp_config: TemporaryConfiguration,
+                           monkeypatch: pytest.MonkeyPatch) -> None:
     import papis.config
 
     general_name = papis.config.get_general_settings_name()
@@ -178,6 +191,7 @@ def test_set_lib_from_path(tmp_config: TemporaryConfiguration) -> None:
     import papis.config
 
     assert tmp_config.libdir is not None
+
     papis.config.set_lib_from_name(tmp_config.libdir)
     assert papis.config.get_lib_name() == tmp_config.libdir
 
@@ -186,7 +200,9 @@ def test_set_lib_from_real_lib(tmp_config: TemporaryConfiguration) -> None:
     import papis.config
 
     libname = "test-set-lib"
-    papis.config.set("dir", tmp_config.libdir, section=libname)
+    papis.config.set("dir",
+                     papis.config.escape_interp(tmp_config.libdir),
+                     section=libname)
 
     assert tmp_config.libdir is not None
     assert os.path.exists(tmp_config.libdir)

--- a/tests/test_utils.py
+++ b/tests/test_utils.py
@@ -13,7 +13,7 @@ def test_get_cache_home(tmp_config: TemporaryConfiguration, monkeypatch) -> None
     assert get_cache_home() == os.path.join(tmp_config.tmpdir, "papis")
     with tempfile.TemporaryDirectory(dir=tmp_config.tmpdir) as d:
         tmp = os.path.join(d, "blah")
-        papis.config.set("cache-dir", tmp)
+        papis.config.set("cache-dir", papis.config.escape_interp(tmp))
         assert get_cache_home() == tmp
 
 

